### ISSUE TITLE
Support configuration for git diff?

### ISSUE THREAD

**Issue by rayluo** [jesseduffield/lazygit#800]
**Is your feature request related to a problem? Please describe.**
There is some fancy [usage for git diff](https://blog.ipspace.net/2020/04/git-tip-word-diff.html). Can we have a configuration option for git diff?

**Describe the solution you'd like**
I apologize if there is already one of such settings available, I'm just not seeing one when examining #797. The current default config file is largely empty, by the way. Could/shall we add all available options there, while commenting out, so that users can easily find and toggle them? - Umm, this alone could be a separated feature requirement.

**UPDATE 2025**
Starting from upcoming lazygit 0.55, you will be able to opt in for git's existing external diff setup by configuring:

```yaml
git:
  paging:
    # If true, Lazygit will use git's `diff.external` config for paging. The advantage over `externalDiffCommand` is that this can be configured per file type in .gitattributes; see https://git-scm.com/docs/gitattributes#_defining_an_external_diff_driver.
    useExternalDiffGitConfig: true
```

**Comment by ahmadie** [comment]
I also would like to use 
git diff --ignore-space-change 
which make things more readable regarding logic specially after applying some formatting

**Comment by ahmadie** [comment]
will i found workaround add following to config yaml

git:
  paging:
    colorArg: always -w
 
then replace [-w] with whatever command u like to be passed to diff command, becareful bcz colorArg is also used for that, [-w] will be passed to stash show -w and git show -w, I hope it does not cause unintended bugs

**Comment by rayluo** [comment]
Oh, since posting this enhancement request, I also asked around, got [help from a StackOverflow Q&A](https://stackoverflow.com/questions/2788649/is-it-possible-to-have-all-git-diff-commands-use-the-python-diff-in-all-git/18948381#18948381), and finalized a self-contained solution OUTSIDE of lazygit. That means, the customized diff effect will work, regardless of whether you are using lazygit or just the good old `git` itself. All the pieces you will need are in my [this commit](https://github.com/rayluo/home/commit/36fb51a886d747e811cc64edd0a2c71cde69fe40) (which is part of [my own dotfile repo](https://github.com/rayluo/home)).

**Comment by jesseduffield** [comment]
Is this something we could achieve by reading the user's gitconfig? Or is the setting typically stored in some other way? If the latter I'm thinking we make it a config flag inside lazygit's own config

**Comment by rayluo** [comment]
> Is this something we could achieve by reading the user's gitconfig?

My understanding is, the underlying git will already pick up the user's gitconfig, as it should. So, lazygit does NOT need to do anything in this case.

> Or is the setting typically stored in some other way? If the latter I'm thinking we make it a config flag inside lazygit's own config

The proposal of this feature request was to see if lazygit would decide to pre-configure [the configuration](https://github.com/rayluo/home/commit/36fb51a886d747e811cc64edd0a2c71cde69fe40) out-of-the-box.

**Comment by jesseduffield** [comment]
@ahmadie, not directly related to this issue, but we now allow the `--ignore-all-space` flag to git diff from inside the app with the <c-w> key

**Comment by rayluo** [comment]
At May 21st, 2020, I wrote,

> Oh, since posting this enhancement request, I also asked around, got [help from a StackOverflow Q&A](https://stackoverflow.com/questions/2788649/is-it-possible-to-have-all-git-diff-commands-use-the-python-diff-in-all-git/18948381#18948381), and finalized a self-contained solution OUTSIDE of lazygit. That means, the customized diff effect will work, regardless of whether you are using lazygit or just the good old `git` itself. All the pieces you will need are in my [this commit](https://github.com/rayluo/home/commit/36fb51a886d747e811cc64edd0a2c71cde69fe40) (which is part of [my own dotfile repo](https://github.com/rayluo/home)).

Recently, @tienlonghungson brought my attention back to this old issue. Now I realized that, the [lazygit 0.20.4](https://github.com/jesseduffield/lazygit/releases/tag/v0.20.4) released at the time when I posted the message above, was ABLE to AUTOMATICALLY utilize - or at least let the native git utilize - [my settings in `~/.config/git/attributes` and `~/.gitconfig`](https://github.com/rayluo/home/commit/36fb51a886d747e811cc64edd0a2c71cde69fe40). Somehow that behavior has long gone in recent versions of lazygit, even after I explicitly config lazygit 0.54.2 to [use git config](https://github.com/jesseduffield/lazygit/blob/master/docs/Custom_Pagers.md#using-git-config).

@stefanhaller , you mentioned something like "adding new keybindings for all of the new behaviors doesn't scale well" in [your another comment in #3984](https://github.com/jesseduffield/lazygit/issues/3984#issuecomment-2409003400). However, in the lazygit 0.20 era, we did not even need any keybinding for this behavior. Lazygit was capable to automatically let git just follow git's config.

Is this a regression introduced between then and now?

**Comment by stefanhaller** [comment]
@rayluo I can have a look if I find some time, but it might take a while (not the most important thing for me to spend time on right now). Feel free to ping me again if I forget.

Just to answer @fredcallaway's comment above: I understand the need, but I feel that using a custom pager that can do intra-line highlighting of changes is a better solution than word-diff. For example, look at [this recent commit](https://github.com/jesseduffield/lazygit/commit/985013718d0f3b246c65a9e4c84508ab0bd8aa6d) (the github UI does a terrible job at rendering it). With `--word-diff` it looks like this: 
<img width="780" height="414" alt="Image" src="https://github.com/user-attachments/assets/15cb3ef3-e1bb-4029-aee3-71de6e1151c4" />

This is not great, it's still next to impossible to see what the change is.

With delta it looks like this, which I find much, much better:
<img width="873" height="705" alt="Image" src="https://github.com/user-attachments/assets/f50c45c4-4981-4ce7-819a-f7ca7ab16022" />

**Comment by rayluo** [comment]
> At May 21st, 2020, I wrote,
> 
> > Oh, since posting this enhancement request, I also asked around, got [help from a StackOverflow Q&A](https://stackoverflow.com/questions/2788649/is-it-possible-to-have-all-git-diff-commands-use-the-python-diff-in-all-git/18948381#18948381), and finalized a self-contained solution OUTSIDE of lazygit. That means, the customized diff effect will work, regardless of whether you are using lazygit or just the good old `git` itself. All the pieces you will need are in my [this commit](https://github.com/rayluo/home/commit/36fb51a886d747e811cc64edd0a2c71cde69fe40) (which is part of [my own dotfile repo](https://github.com/rayluo/home)).
> 
> ... Now I realized that, the [lazygit 0.20.4](https://github.com/jesseduffield/lazygit/releases/tag/v0.20.4) released at the time when I posted the message above, was ABLE to AUTOMATICALLY utilize ... [my settings in `~/.config/git/attributes` and `~/.gitconfig`](https://github.com/rayluo/home/commit/36fb51a886d747e811cc64edd0a2c71cde69fe40). Somehow that behavior has long gone in recent versions of lazygit, even after I explicitly config lazygit 0.54.2 to [use git config](https://github.com/jesseduffield/lazygit/blob/master/docs/Custom_Pagers.md#using-git-config).
> 
> ......
> 
> Is this a regression introduced between then and now?

@stefanhaller , I investigated and identified [this change between lazygit 0.22.9 and 0.23.1](https://github.com/jesseduffield/lazygit/pull/1012) caused the regression. The change was used to [bypass the external diff tool (because in that case it was a GUI tool bcompare)](https://github.com/jesseduffield/lazygit/issues/1010). However, in my case I am using a console-based diff script, so, it is still desirable to call "git diff" without the "--no-ext-diff" flag. How shall we modify lazygit to support both legit scenarios? Shall we introduce a new param like this:

```yaml
git:
  paging:
    diffArgs: --no-ext-diff
```

so that people can opt in? The default value shall be empty so that the default behavior follows whatever existing behavior that is already defined in git. Doesn't that make sense?

**Comment by stefanhaller** [comment]
I'm a little confused here; I'm not sure I understand why your approach worked before 0.22.9. When I just tested custom diff drivers (defined in .gitattributes), I found that they are only used when you also pass `--ext-diff` to the `git diff` (or `git show`) call. We didn't do this in lazygit 0.22.9, so why did this work?

(This is not very well documented in git, I feel. It isn't mentioned in the .gitattributes manpage at all, but it is mentioned in the docs for the `--ext-diff` option, albeit only for `git log`.)

**Comment by rayluo** [comment]
> ... I'm not sure I understand why your approach worked before 0.22.9. When I just tested custom diff drivers (defined in .gitattributes), I found that they are only used when you also pass `--ext-diff` to the `git diff` (or `git show`) call. We didn't do this in lazygit 0.22.9, so why did this work?

Wait, was there a typo on the `--ext-diff` in your message? My understanding is that [lazygit started to pass `--no-ext-diff` sine 0.23, but never passes `--ext-diff`](https://github.com/jesseduffield/lazygit/pull/1012/files#diff-2f099b446f458f951858bb0412dd630dbed34fdf961cc4b72fac203f2a073162L791-R791).
>
> (This is not very well documented in git, I feel. It isn't mentioned in the .gitattributes manpage at all, but it is mentioned in the docs for the --ext-diff option, albeit only for git log.)

Agreed that [the docs](https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---ext-diff) is too vague. They have a pair of opposite parameters `--ext-diff` and `--no-ext-diff`, but did not mention which behavior is the default. It is up to anyone's interpretation now. Could it be that the `--ext-diff` is only necessary when using "git log ..."? (But then I don't even know what that means. I never saw any diff content when I use "git log", even after I tentatively try "git log --ext-diff" just now.)

Regardless, my observation is that [my usage](https://github.com/rayluo/home/commit/36fb51a886d747e811cc64edd0a2c71cde69fe40#diff-3cbe50069f7ffe4d55b4d1fe7a72fb79e9c9545af55d8326a728b98f00481bf2) has always been working outside of lazygit, i.e. I can do a good old "git diff" in terminal and see my customized diff effect. The lazygit at that era (<=0.22.9) simply allowed that to happen. Was it because lazygit <= 0.22.9 did not use `--ext-diff` nor `--no-ext-diff`, so the git's default behavior will be preserved?

**Comment by stefanhaller** [comment]
> > ... I'm not sure I understand why your approach worked before 0.22.9. When I just tested custom diff drivers (defined in .gitattributes), I found that they are only used when you also pass `--ext-diff` to the `git diff` (or `git show`) call. We didn't do this in lazygit 0.22.9, so why did this work?
> 
> My understanding is that lazygit started to pass `--no-ext-diff` sine 0.23, but never passes `--ext-diff`.

Yes, that was my point. No typo here, we didn't pass either `--ext-diff` or `--no-ext-diff`, so it shouldn't have worked, because you need to pass `--ext-diff` in order to get the custom diff driver to be used. Or so I thought.

It turns out that you need to pass it only for `git show`, but not for `git diff`. Which means that your custom approach should have worked only in the Files panel, but not in the Commits panel, can you confirm that? (I'm too lazy to test it myself 😉)

> I never saw any diff content when I use "git log", 

You can use `-p` with `git log`, in which case it includes the diff in the log.

**Comment by rayluo** [comment]
> you need to pass `--ext-diff` in order to get the custom diff driver to be used. Or so I thought.
> 
> It turns out that you need to pass it only for `git show`, but not for `git diff`. Which means that your custom approach should have worked only in the Files panel, but not in the Commits panel, can you confirm that?

Yes, I tested it just now with lazygit 0.22.9, and I can confirm that is indeed the case. Now that you brought it up, I think it would be nice if the Commits panel can also follow the same diff behavior of Files panel. But at least please get the Files panel right. :-)



> > I never saw any diff content when I use "git log",
> 
> You can use `-p` with `git log`, in which case it includes the diff in the log.

Good to know. Thanks!

**Comment by stefanhaller** [comment]
@rayluo Here's a PR: #4832. Please review/test!

I made this a draft PR for now because I first want to make another PR that removes the awkward `Git.Paging.UseConfig` option, which is becoming even more confusing with this change.

**Comment by stefanhaller** [comment]
@rayluo Do you think we can close this, now that #4832 is merged?

**Comment by rayluo** [comment]
Yes, feel free to close this issue. Meanwhile, I am appending this to the issue description:

Starting from upcoming lazygit 0.55, you will be able to opt in for git's existing external diff setup by configuring:

```yaml
git:
  paging:
    # If true, Lazygit will use git's `diff.external` config for paging. The advantage over `externalDiffCommand` is that this can be configured per file type in .gitattributes; see https://git-scm.com/docs/gitattributes#_defining_an_external_diff_driver.
    useExternalDiffGitConfig: true
```

By the way, that "# If true ..." comment [looked OK in the PR](https://github.com/jesseduffield/lazygit/pull/4832#diff-798caf80d294e3e03a8bb98eed6140be9dbbb41fe712ce59ccf6f1a8ced89d24R309), but it is too long when [rendered after merging](https://github.com/jesseduffield/lazygit/blob/master/docs/Config.md). Not sure whether you would want to add another commit to split it into multiple lines.

**Comment by stefanhaller** [comment]
> By the way, that "# If true ..." comment [looked OK in the PR](https://github.com/jesseduffield/lazygit/pull/4832#diff-798caf80d294e3e03a8bb98eed6140be9dbbb41fe712ce59ccf6f1a8ced89d24R309), but it is too long when [rendered after merging](https://github.com/jesseduffield/lazygit/blob/master/docs/Config.md). Not sure whether you would want to add another commit to split it into multiple lines.

This is consistent with all the other comments in the file; we have lots of very long, single-line comments. Apparently the idea was that these are paragraphs that get auto-wrapped during markdown rendering, which of course doesn't happen in a fenced code block. I'll put this on the list to be cleaned up at some point, thanks for mentioning it.
